personalities/sctfe/storage/bbolt/dedup.go

```diff
@@ -0,0 +1,179 @@
+		return nil, fmt.Errorf("bolt.Open(): %v", err)
+	}
+	s := &Storage{db: db}
+
+	err = db.Update(func(tx *bolt.Tx) error {
+		dedupB := tx.Bucket([]byte(dedupBucket))
+		sizeB := tx.Bucket([]byte(sizeBucket))
+		if dedupB == nil && sizeB == nil {
+			klog.V(2).Infof("NewStorage: no pre-existing buckets, will create %q and %q.", dedupBucket, sizeBucket)
+			_, err := tx.CreateBucket([]byte(dedupBucket))
+			if err != nil {
+				return fmt.Errorf("create %q bucket: %v", dedupBucket, err)
+			}
+			sb, err := tx.CreateBucket([]byte(sizeBucket))
+			if err != nil {
+				return fmt.Errorf("create %q bucket: %v", sizeBucket, err)
+			}
+			klog.V(2).Infof("NewStorage: initializing %q with size 0.", sizeBucket)
+			err = sb.Put([]byte("size"), itob(0))
+			if err != nil {
+				return fmt.Errorf("error reading logsize: %v", err)
+			}
+		} else if dedupB == nil && sizeB != nil {
+			return fmt.Errorf("inconsistent deduplication storage state %q is nil but %q it not nil", dedupBucket, sizeBucket)
+		} else if dedupB != nil && sizeB == nil {
+			return fmt.Errorf("inconsistent deduplication storage state, %q is not nil but %q is nil", dedupBucket, sizeBucket)
+		} else {
+			klog.V(2).Infof("NewStorage: found pre-existing %q and %q buckets.", dedupBucket, sizeBucket)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("error initializing buckets: %v", err)
+	}
+
+	return s, nil
+}
+
+// Add inserts entries in the dedup bucket and updates the size bucket if need be.
+//
+// If an entry is already stored under a given key, Add does not update it, even if the new value
+// is different.
+// The context is here for consistency with interfaces, but isn't used by BBolt.
+func (s *Storage) Add(_ context.Context, kvs []dedup.KV) error {
+	for _, kv := range kvs {
+		err := s.db.Update(func(tx *bolt.Tx) error {
+			db := tx.Bucket([]byte(dedupBucket))
+			sb := tx.Bucket([]byte(sizeBucket))
+			sizeB := sb.Get([]byte("size"))
+			if sizeB == nil {
+				return fmt.Errorf("can't find log size in bucket %q", sizeBucket)
+			}
+			size := btoi(sizeB)
+
+			if old := db.Get(kv.K); old != nil {
+				klog.V(3).Infof("Add(): bucket %q already contains an entry for %q, not updating", dedupBucket, hex.EncodeToString(kv.K))
+			}
+			if err := db.Put(kv.K, itob(kv.V)); err != nil {
+				return err
+			}
+			// sizeB is indexes from 1 since it's a size, kv.K from 0.
+			// Therefore, if they're equal, kv is a new entry.
+			if size == kv.V {
+				klog.V(3).Infof("Add(): updating deduped size to %d", size+1)
+				if err := sb.Put([]byte("size"), itob(size+1)); err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("b.Put(): error writting leaf index %d: err", kv.V)
+		}
+	}
+	return nil
+}
+
+// Get reads entries from the dedup bucket.
+//
+// The context is here for consistency with interfaces, but isn't used by BBolt.
+func (s *Storage) Get(_ context.Context, leafID []byte) (uint64, bool, error) {
```

Comment:
The meaning of the returned `bool` is not obvious. Adding the doc comment or naming the returned variables will help.

Reply:
Done. It's using a go idiom called the "comma ok": https://go.dev/doc/effective_go#maps.

Reply:
Correct, this is how the `sync.Map` implements the named return variables with "comma ok".

```go
func (m *Map) Load(key any) (value any, ok bool) {
```

https://github.com/golang/go/blob/fc9f02c7aec81bcfcc95434d2529e0bb0bc03d66/src/sync/map.go#L123